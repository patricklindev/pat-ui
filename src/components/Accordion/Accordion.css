.accordion {
  margin: 0;
  padding: 0;
  list-style: none;
}

.accordion-item {
  border-bottom: 1px solid #ccc;
}

.accordion-item:not(:first-of-type) {
  border-top: 0;
}

.accordion-item-title {
  width: 100%;
  margin: 0;
}

.accordion-item-btn {
  display: flex;
  align-items: center;
  width: 100%;
  background-color: #fff;
  border: 0;
  padding: 15px 20px;
  font-size: 16px;
  font-weight: 400;
  cursor: pointer;
  user-select: none;
}

.accordion-item-btn img {
  width: 16px;
  height: 16px;
  margin-left: auto;
  transition: transform 0.2s ease-in-out;
}

.accordion-item.active {
  margin-bottom: 1rem;
}

.accordion-item.active:not(:first-of-type) {
  margin-top: 1rem;
}

.accordion-item.active img {
  transform: rotate(-180deg);
}

.accordion-item-container {
  transition: height 0.2s ease-in-out;
  overflow: hidden;
}

.accordion-item-content {
  /* border-top: 1px solid #cccccc; */
  padding: 15px 20px;
  background-color: #fff;
}

.accordion-item-btn:disabled {
  background-color: rgba(0, 0, 0, 0.12);
}

.container {
  background-color: #e7ebf0;
  width: 600px;
  max-width: 100%;
  margin: 0 auto;
  padding: 20px 15px;
}
