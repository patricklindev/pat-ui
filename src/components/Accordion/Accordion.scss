.content.show {
  opacity: 1;
  transition: all 5s cubic-bezier(0,1,0,1);
  max-height: 500px;
}
.content.hidden {
  opacity: 0;
  transition: all .5s cubic-bezier(0,1,0,1);
  max-height: 0px;
}

.title {
  cursor: pointer;
  display: flex;
  justify-content: space-between;
  padding: 5px;
}
.item {
  margin: 5px auto;
  padding: 5px;
  box-shadow: 1px 2px 1px gray ;
  border: 1px solid gray;
}
h2 {
  font-size: 20px;
  align-self: center;
  margin: 0;
}
span {
  padding: 10px;
}
.arrow-up {
  -webkit-transform:rotate(180deg);
}
