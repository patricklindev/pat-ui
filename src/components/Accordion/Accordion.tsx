import { AccordionData } from './types';
import AccordionItem from './AccordionItem';
import { useState } from 'react';
import React from 'react';

function Accordion({ items }: { items: Array<AccordionData> }) {
	const [currentIdx, setCurrentIdx] = useState(-1);
	
	const btnOnClick = (idx: number) => {
		setCurrentIdx((currentValue) => (currentValue !== idx ? idx : -1));
	      };
	      
	return (
	  <ul className="accordion">
	    {items.map((item, idx) => (
	      <AccordionItem
		key={idx}
		data={item}
		isOpen={idx === currentIdx}
		btnOnClick={() => btnOnClick(idx)}
	      />
	    ))}
	  </ul>
	);
      }

export default Accordion;