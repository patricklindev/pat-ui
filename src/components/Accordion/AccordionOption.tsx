import React, { ReactElement, ReactNode, useEffect, useState } from "react"; 
import Icon from "../Icon";
import { IIconProps } from "../Icon/Icon";

export interface IAccordionOption {
    summary: ReactNode;
    detail: ReactNode;
    expandIcon?: ReactElement<IIconProps>;
    expandProp?: boolean;
    onChange?: ()=>void;
}

export default function AccordionOption({summary="", detail="", expandIcon=<Icon name="angle down"/>, expandProp=false, onChange=()=>{}}: IAccordionOption):JSX.Element {
    const [expanded, setExpanded] = useState<boolean>(expandProp)

    useEffect(()=> {
        onChange()
    } ,[expanded])

    function onToggle(e: React.MouseEvent) {
        setExpanded(!expanded)
    }

    // expanded ? "accordion__expanded": "">
    return <div className={(expanded ? "accordionOption-expanded" : "accordionOption")}>
        <div onClick={onToggle} className={(expanded ? "accordionOption__summary-bar-expanded" : "accordionOption__summary-bar")}>
            <div className="accordionOption__summary">{summary}</div>
            <div role="icon" className="accordionOption__expandIcon">{expandIcon}</div>
        </div>
        <div role="detail" className="accordionOption__detail" aria-expanded={!expanded}>{detail}</div>
    </div>
}
