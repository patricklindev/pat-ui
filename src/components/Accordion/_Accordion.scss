.accordion {
    box-shadow: 0 0.1em $gray-300;
}
.accordionOption {
    background-color: $white;
    margin-bottom: 1px;
    padding: 0.75em 1em;
    border-bottom: 0.1em solid $gray-500;
}

.accordionOption-expanded {
    background-color:$white;
    margin-bottom: 1em;
    padding: 1.25em 1em 1em;
    border-bottom: 0.1em solid $gray-500;
}

.accordionOption__summary-bar {
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: inherit;
}

.accordionOption__summary-bar-expanded {
    display: flex;
    justify-content: space-between;
    align-items: center;
    background-color: inherit;
    margin-bottom: 1.75em;
}

.accordionOption__summary {
    background-color: inherit;
}

.accordionOption__expandIcon {
    background-color: inherit;
}

.accordionOption__detail[aria-expanded="true"] {
    background-color: inherit;
    max-height: 0px;
    overflow: hidden;
    transition: max-height 0.5s cubic-bezier(0, 1, 0, 1);
}

.accordionOption__detail {
    background-color: inherit;
    max-height: 2000px;
    overflow: hidden;
    transition: max-height 1s ease-in-out;
}

