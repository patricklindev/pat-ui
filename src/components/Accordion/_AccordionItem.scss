.accordion-item {
	border: 1px solid #ccc;
      }
      
      .accordion-item:not(:first-of-type) {
	border-top: 0;
      }
      
      .accordion-item-title {
	width: 100%;
	margin: 0;
	// text-align: left;
      }
      
      .accordion-item-btn {
	display: flex;
	align-items: center;
	width: 100%;
	background-color: #fff;
	border: 0;
	padding: 15px 20px;
	font-size: 18px;
	font-weight: 400;
	cursor: pointer;
	user-select: none;
      }
      
      .accordion-item-btn::after {
	content: '';
	flex-shrink: 0;
	width: 18px;
	height: 18px;
	margin-left: auto;
	background-image: url("data:image/svg+xml,%3csvg xmlns='http://www.w3.org/2000/svg' viewBox='0 0 16 16' fill='%23212529'%3e%3cpath fill-rule='evenodd' d='M1.646 4.646a.5.5 0 0 1 .708 0L8 10.293l5.646-5.647a.5.5 0 0 1 .708.708l-6 6a.5.5 0 0 1-.708 0l-6-6a.5.5 0 0 1 0-.708z'/%3e%3c/svg%3e");
	background-repeat: no-repeat;
	background-size: 18px;
	transition: transform 0.2s ease-in-out;
      }
      
      .accordion-item.active .accordion-item-btn::after {
	transform: rotate(-180deg);
      }
      
      .accordion-item-container {
	transition: height 0.2s ease-in-out;
	overflow: hidden;
      }
      
      .accordion-item-content {
	border-top: 1px solid #cccccc;
	padding: 15px 20px;
      }
      .title {
	
    	padding:0;
    	margin:0;
      }
      .subTitle {
	
	color: #cccccc;
	padding-left: 150px;
      }	
      #square {
	width: 30px;
      }
      .accordion {
	list-style: none;
	margin: 0;
	padding: 0;
      }
      .container {
	width: 600px;
	max-width: 100%;
	margin: 0 auto;
	padding: 20px 15px;
      } 
      * {
	box-sizing: border-box;
      }
      
      body {
	margin: 0;
	font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto',
	  'Oxygen', 'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans',
	  'Helvetica Neue', sans-serif;
	-webkit-font-smoothing: antialiased;
	-moz-osx-font-smoothing: grayscale;
	line-height: 1.6;
      }
      
      