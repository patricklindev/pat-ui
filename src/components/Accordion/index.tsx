export {default as Accordion} from "./Accoridion"
export {default as AccordionOption} from "./AccordionOption"