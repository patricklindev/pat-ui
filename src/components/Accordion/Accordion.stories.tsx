import React from 'react';
import { action } from '@storybook/addon-actions';
import AccordionOption from './AccordionOption';
import Accordion from './Accoridion';

export default {
  title: 'Accordion',
  component: Accordion,
};

const accordionStyle: React.CSSProperties = {
  marginRight: '5px',
  marginTop: '5px',
};

export const DefaultAccordion = () => (
    <Accordion>
        <AccordionOption
        summary="default summary" 
        detail="You can place a string, React component or a JSX Element"
        onChange={action("expanded component")} />
        <AccordionOption
        summary="default summary" 
        detail="You can place a string, React component or a JSX Element"
        onChange={action("expanded component")} />
    </Accordion>
);
