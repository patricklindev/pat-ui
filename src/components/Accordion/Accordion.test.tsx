import { fireEvent, render } from "@testing-library/react";
import React from "react";
import Icon from "../Icon";
import AccordionOption from "./AccordionOption";
import Accordion from "./Accoridion";

describe("Accordion", ()=> {
    it("should match snapshot", ()=>{
        const {asFragment} = render(
            <Accordion>
                <AccordionOption
                summary="SnapShot"
                detail="SnapShot"
                />
            </Accordion>
        )
        expect(asFragment()).toMatchSnapshot();
    })

    it("should render summary and expandIcon", ()=>{
        const wrapper = render(
            <Accordion>
                <AccordionOption
                summary="summary"
                detail="detail"
                />
            </Accordion>
        )
        expect(wrapper.queryByText("summary")).toBeInTheDocument();
        expect(wrapper.queryByRole("icon")).toBeInTheDocument();
    })
    
    it("should render detail", ()=> {
        const wrapper = render(
            <Accordion>
                <AccordionOption
                summary="summary"
                detail="detail"
                expandProp={true}
                />
            </Accordion>
        )
        expect(wrapper.queryByText("detail")).toHaveAttribute("aria-expanded", "false");
    })

    it("should not render detail", ()=> {
        const wrapper = render(
            <Accordion>
                <AccordionOption
                summary="summary"
                detail="detail"
                expandProp={false}
                />
            </Accordion>
        )
        expect(wrapper.queryByText("detail")).not.toHaveAttribute("aria-expanded", "false");
    })

    it("call back function should be called", ()=>{
        const defaultProps = {
            summary: "summary",
            detail: "detail",
            expandProp: false,
            onChange: jest.fn() 
        }
        const wrapper = render(
            <Accordion>
                <AccordionOption {...defaultProps}/>
            </Accordion>
        )
        expect(defaultProps.onChange).toHaveBeenCalledTimes(1);
        fireEvent.click(wrapper.getByText("summary"));
        expect(defaultProps.onChange).toHaveBeenCalledTimes(2);
    })

    it("clicking should show detail component", ()=>{
        const wrapper = render(
            <Accordion>
                <AccordionOption
                summary="summary"
                detail="detail"
                expandProp={false}
                />
            </Accordion>
        )
        fireEvent.click(wrapper.getByText("summary"))
        expect(wrapper.queryByText("detail")).toHaveAttribute("aria-expanded", "false");
    })

    it("clicking should close detail component", ()=>{
        const wrapper = render(
            <Accordion>
                <AccordionOption
                summary="summary"
                detail="detail"
                expandProp={true}
                />
            </Accordion>
        )
        fireEvent.click(wrapper.getByText("summary"))
        expect(wrapper.queryByText("detail")).toHaveAttribute("aria-expanded", "true");
    })

    it("clicking the component multiple time works", ()=>{
        const wrapper = render(
            <Accordion>
                <AccordionOption
                summary="summary"
                detail="detail"
                expandProp={false}
                />
            </Accordion>
        )
        let expanded = false
        for (let i = 0; i < 10; i++) {
            fireEvent.click(wrapper.getByText("summary"))
            expanded = !expanded
            if (expanded) {
                expect(wrapper.queryByText("detail")).toHaveAttribute("aria-expanded", "false");
            } else {
                expect(wrapper.queryByText("detail")).toHaveAttribute("aria-expanded", "true");
            }
        }
    })
})