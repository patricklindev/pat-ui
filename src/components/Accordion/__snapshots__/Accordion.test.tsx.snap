// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="accordion"
  >
    <div
      class="accordionOption"
    >
      <div
        class="accordionOption__summary-bar"
      >
        <div
          class="accordionOption__summary"
        >
          SnapShot
        </div>
        <div
          class="accordionOption__expandIcon"
          role="icon"
        >
          <svg
            class="icon angle down"
            height="28"
            viewBox="0 0 320 512"
          >
            <path
              d="M143 352.3L7 216.3c-9.4-9.4-9.4-24.6 0-33.9l22.6-22.6c9.4-9.4 24.6-9.4 33.9 0l96.4 96.4 96.4-96.4c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9l-136 136c-9.2 9.4-24.4 9.4-33.8 0z"
            />
          </svg>
        </div>
      </div>
      <div
        aria-expanded="true"
        class="accordionOption__detail"
        role="detail"
      >
        SnapShot
      </div>
    </div>
  </div>
</DocumentFragment>
`;
